### DISCUSSION TITLE
Custom column in mitmweb for flow number (iterates from 1 to n)

### DISCUSSION THREAD

**Discussion by malwareinfosec** [mitmproxy/mitmproxy#6976] [Q&A] [answered]
Hi, I'm looking for a friendly way to refer to a flow in mitmweb. Flow.id is not really meaningful, and while the ``timestamp`` column is great, I'm looking for something numerical starting from 1 to n. For example, something like this:
![image](https://github.com/mitmproxy/mitmproxy/assets/25351665/831d3874-c36f-4328-9c56-ca8e4dee9801)

Is this already built-in by any chance?

**Comment by mhils** [comment]
No, there isn't right now. The problem with flow number is that it changes (users may delete individual flows, or they may duplicate an existing flow), so that's not ideal when referring to individual flows. What's your specific use case?

**Reply by malwareinfosec** [reply]
Thanks Max. My use case is to easily refer to a specific flow in the context of my addon. For example:

- identifying which flow had a search term match within the response body (this is related to this [thread](https://github.com/mitmproxy/mitmproxy/discussions/6973#discussioncomment-9908801)):
![image](https://github.com/mitmproxy/mitmproxy/assets/25351665/a10bac38-ef3c-4b67-a372-4aa1b687bfe9)
In the image above you see that I am using the timestamp as a way to pinpoint which flow it is.  I am also using ```flow.marked = ":red:"``` and ```flow.comment = 'Found'```, but Instead, I would prefer something like: ```Found string in flow #15```. This would allow you to scroll by ```flow #``` within the UI without the need to do a filter by ```~marked```.

- using a flow number to process certain queries. For example, let's say your starting point is at ```flow #10```. You could then do a search within ```flow #1 to #9``` specifically. I suppose you could also use epoch time and do something like ```< starting_epochtime``` but it's not as friendly.

As far as handling the scenarios you mentioned (user deletes or duplicates flows), here's how other similar applications handle it:
![image](https://github.com/mitmproxy/mitmproxy/assets/25351665/16d08d36-94ed-43bd-a7e1-bf442d0c3bf9)
Essentially, the ```flow #``` keeps on being incremented and deleted values are not replaced. The only case where the counter is reset to ```flow #1``` is when a user clears all flows.

**Reply by malwareinfosec** [reply]
@mhils revisiting this question to ask if the indexing that currently exists in mitmproxy CLI can be accessed in mitmweb (and potentially added as a column):

![image](https://github.com/user-attachments/assets/1eab46c8-3341-4985-8182-b4a02877a688)

**Reply by mhils** [reply]
Getting a "simple" index is easy, we do have a `flows.listIndex` in the web ui state which maps from uuid to the current index. Making it so that deleted values are not replaced would require significant more effort, and may not cross the usefulness vs maintenance effort threshold for me.

**Reply by malwareinfosec** [reply]
How are deleted values currently handled in CLI? From what I can see deleting flows, the next flow inherits the index value from the deleted flow. This is acceptable with the understanding that the index is not a static unique identifier. It still serves its purpose as a quick way to jump to a flow by knowing its index.

**Reply by mhils** [reply]
> How are deleted values currently handled in CLI? 

It's basically the list index for all current flows.

**Reply by malwareinfosec** [reply]
OK, thanks. Can that list index be displayed in column via ``loader.add_option``?

![image](https://github.com/user-attachments/assets/fcf71311-e0dc-4cd5-af70-93ddffc28098)

**Reply by mhils** [reply, marked as answer]
#7039 adds an index column, which can then be loaded with the `web_columns` option. :)

(When using mitmproxy from GitHub you need to update the precompiled web assets, see `web/README.md`. They will be updated with the next release)

**Reply by mhils** [reply]
FYI, 10.4 is out. :)

**Reply by malwareinfosec** [reply]
Nice! One question, I'm trying to access the value of the index via addon, using ```flow.listIndex```, but it's not returning anything. Any idea?

![image](https://github.com/user-attachments/assets/f4b00a0c-13f1-4e08-bd08-418d6c5def6a)

**Reply by mhils** [reply]
The following works:

```python
from mitmproxy.ctx import master

def request(flow):
    print(master.view.index(flow))
```

This is using an internal API so there are no guarantees, but you should be reasonably fine here. :)

**Reply by malwareinfosec** [reply]
Noticed something with the UI, and not sure if it is by design or not. You cannot reorder the flow index like you can for the other columns (clicking on the top column and reorder ascending/descending)

![image](https://github.com/user-attachments/assets/670e11ad-90af-4986-a305-d7edb0a88d92)

**Reply by mhils** [reply]
https://github.com/mitmproxy/mitmproxy/pull/7039#discussion_r1690833560 - something we can hopefully fix eventually, but for now that's a known limitation.
